src/fuzz_introspector/frontends/src/frontend_go.py

```diff
@@ -0,0 +1,610 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+################################################################################
+"""Fuzz Introspector Light frontend for Go"""
+
+from typing import Optional
+
+import os
+import pathlib
+import re
+
+import logging
+
+from tree_sitter import Language, Parser, Node
+import tree_sitter_go
+import yaml
+
+logger = logging.getLogger(name=__name__)
+LOG_FMT = '%(asctime)s.%(msecs)03d %(levelname)s %(module)s - %(funcName)s: %(message)s'
+
+
+class SourceCodeFile():
+    """Class for holding file-specific information."""
+
+    def __init__(self,
+                 source_file: str,
+                 source_content: Optional[bytes] = None):
+        logger.info('Processing %s' % source_file)
+
+        self.root = None
+        self.imports = []
+        self.source_file = source_file
+        self.tree_sitter_lang = Language(tree_sitter_go.language())
+        self.parser = Parser(self.tree_sitter_lang)
+
+        if source_content:
+            self.source_content = source_content
+        else:
+            with open(self.source_file, 'rb') as f:
+                self.source_content = f.read()
+
+        # List of function definitions in the source file.
+        self.functions = []
+        self.methods = []
+
+        # Initialization ruotines
+        self.load_tree()
+
+        # Load function/method declaration
+        self._set_function_declaration()
+        self._set_method_declaration()
+
+        # Parse import package
+        self._set_imports()
+
+    def load_tree(self):
+        """Load the the source code into a treesitter tree, and set
+        the root node."""
+        self.root = self.parser.parse(self.source_content).root_node
+
+    def _set_function_declaration(self):
+        """Internal helper for retrieving all functions."""
+        func_query_str = '( function_declaration ) @fd '
+        func_query = self.tree_sitter_lang.query(func_query_str)
+
+        function_res = func_query.captures(self.root)
+        for _, funcs in function_res.items():
+            for func in funcs:
+                self.functions.append(
+                    FunctionMethod(func, self.tree_sitter_lang, self, True))
+
+    def _set_method_declaration(self):
+        """Internal helper for retrieving all methods."""
+        func_query_str = '( method_declaration ) @fd '
+        func_query = self.tree_sitter_lang.query(func_query_str)
+
+        function_res = func_query.captures(self.root)
+        for _, funcs in function_res.items():
+            for func in funcs:
+                self.methods.append(
+                    FunctionMethod(func, self.tree_sitter_lang, self, False))
+
+    def _set_imports(self):
+        """Internal helper for retrieving all imports."""
+        import_pattern = r'^\s*import\s+(?:(?P<single>"[^"]+")|\((?P<multi>[^)]+)\))'
```

Comment:
Do we really need to look at the source code for this? Can't we traverse the syntax tree?

Just looked at the tree for simple program:

```go
package main

import "fmt"

import "rsc.io/quote"

func main() {
    fmt.Println(quote.Go())
}
```
gives the tree:
```
source_file [0, 0] - [9, 0]
  package_clause [0, 0] - [0, 12]
    package_identifier [0, 8] - [0, 12]
  import_declaration [2, 0] - [2, 12]
    import_spec [2, 7] - [2, 12]
      path: interpreted_string_literal [2, 7] - [2, 12]
  import_declaration [4, 0] - [4, 21]
    import_spec [4, 7] - [4, 21]
      path: interpreted_string_literal [4, 7] - [4, 21]
      ...
```

We should be able to traverse `import_declaration`?

Sample logic in C: https://github.com/ossf/fuzz-introspector/blob/60cb9940908f1ab6a4744e3e218baafe0393a618/src/fuzz_introspector/frontends/src/frontend_c.py#L490-L503

Reply:
Have changed.